[4.1] Field group site display enhancement

ceford opened the pull request:
Pull Request for Issue #31385 and related to #31396  .

### Summary of Changes

Code to customise site display of Field Groups

### Testing Instructions

Create some article fields that are not in a field group.
Edit an article and populate those fields.

Test the default behaviour.

Create one or more field groups, the example I use is Ingredients and Procedure.
The PR has added some extra fields to the Options tab of the Field Group data entry form:
	Select Show Title = Yes
	Select a render tag.
Add some fields to those groups.
Edit an article and populate the fields

### Actual result BEFORE applying this Pull Request

The default behaviour is to show all fields in a single bulleted list with no 'title'.

### Expected result AFTER applying this Pull Request

The example I use shows a Recipe with Ingredients as a bulleted list and the Procedure as a numbered list, with titles.

![image](https://user-images.githubusercontent.com/1560390/99184978-c61a4c80-273e-11eb-81d3-ad36d887b11b.png)

### Documentation Changes Required

Help screen will need revision

ceford committed:
# WARNING: head commit changed in the meantime

removed two local html files from system

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
editor xml

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

brianteeman commented on administrator/components/com_fields/forms/group.xml:
```diff
@@ -152,6 +152,59 @@
+				label="COM_FIELDS_GROUP_TITLE_TAG_LABEL"
+				default="ul"
+				>
+				<option value="p">P</option>
```
lowercase please

brianteeman commented on administrator/components/com_fields/forms/group.xml:
```diff
@@ -152,6 +152,59 @@
+				label="COM_FIELDS_GROUP_RENDER_TAG_LABEL"
+				default="ul"
+				>
+				<option value="ul">UL</option>
```
lowercase please

brianteeman commented:
Warning: Use of undefined constant none - assumed 'none' (this will throw an Error in a future version of PHP) in C:\htdocs\joomla-cms\components\com_fields\layouts\fields\render.php on line 91

brianteeman commented:
If you dont enter a class then it should not output an empty class. Currently you get for example `<ul class="">`

richard67 commented:
@ceford Please use tabs and not spaces to indent code, see PHP CS errors reported by drone: [https://ci.joomla.org/joomla/joomla-cms/37579/1/6](https://ci.joomla.org/joomla/joomla-cms/37579/1/6)

brianteeman commented:
Depending on the options you select they are either ignored or they generate invalid mark

You cannot have the following markup
<p>title</p>
<li>field1</li>
<li>field2</li>

But it gets much worse with other combinations

brianteeman commented:
Not sure what the show/hide title is supposed to do but I couldnt see it doing anything

brianteeman commented:
I have tested this item :red_circle: unsuccessfully on 29a934a69b55e503458064c334324c4c3559b1cb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31408">issues.joomla.org/tracker/joomla-cms/31408</a>.</sub>

brianteeman commented:
Thanks for the fixes - getting better

![image](https://user-images.githubusercontent.com/1296369/99189909-8b72dd00-275b-11eb-8973-f3906b6776c8.png)

This generates the following

![image](https://user-images.githubusercontent.com/1296369/99189935-ad6c5f80-275b-11eb-8536-0567f5a0f5f8.png)

I get why you made it do that - p and dl would be invalid but I you should not be able to select an invalid option or combination in the first place. Otherwise users will think it is broken

It's the same when you set title to hide on a dl. The title is not hidden because it would be invalid

brianteeman commented on administrator/components/com_fields/forms/group.xml:
```diff
@@ -152,6 +152,59 @@
+				name="title_tag"
+				type="list"
+				label="COM_FIELDS_GROUP_TITLE_TAG_LABEL"
+				default="ul"
```
the default value of `ul` is not present in the list of options?

ceford commented:
Thanks to @brianteeman for obliging me to think about the form behaviour. After wondering how to compose some javascript and poking around I discovered that all I had to do was add a field attribute to the form xml: showon="show_title:1[AND]render_tag:ul,ol,p,div" which I can only describe as absolutely marvelous. My complements to the javascript developers. I wonder if this sort of stuff is documented somewhere?

brianteeman commented:
> I wonder if this sort of stuff is documented somewhere?

https://docs.joomla.org/Form_field#Showon

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ashvini77 commented:
I have tested this item :red_circle: unsuccessfully on 092e829e4dc5051af06c6c71b06ae2cb1e8f5f5e<br /><br />https://tinyurl.com/y5r5ys4z<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31408">issues.joomla.org/tracker/joomla-cms/31408</a>.</sub>

gostn commented:
@ashvini77 you can add a screenshot here so user don't have to open a link.

punambaravkar commented:
I have tested this item :white_check_mark: successfully on 092e829e4dc5051af06c6c71b06ae2cb1e8f5f5e<br /><br />Yes, After adding the patch, it works correctly <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31408">issues.joomla.org/tracker/joomla-cms/31408</a>.</sub>

ashvini77 commented:
![screen shot 2020-12-05 at 07 35 22](https://issues.joomla.org/uploads/1/512a69cf5afc40b81a27f3dd6e2c9ab5.png)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31408">issues.joomla.org/tracker/joomla-cms/31408</a>.</sub>

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

vaibhavsTekdi commented:
I have tested this item :white_check_mark: successfully on 092e829e4dc5051af06c6c71b06ae2cb1e8f5f5e<br /><br />It works only if when I save that field group after applying the patch again.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31408">issues.joomla.org/tracker/joomla-cms/31408</a>.</sub>

ceford commented:
The bug occurs when you set up the test groups before applying the patch. To fix it, apply the patch and then open each Field Group in turn and save. That updates the parameters. I have change the code for backward compatibility purposes. You still need to save the Field Groups with the new features to store them. Please test again.

Quy commented on components/com_fields/layouts/fields/render.php:
```diff
@@ -69,14 +73,68 @@
 
-	$output[] = '<li class="field-entry ' . $class . '">' . $content . '</li>';
+	// check gparams for backward compatibility
+	$gparams = json_decode($field->group_params);
+	
+	if (isset($field->group_title) && isset($gparams->render_tag))
```
```suggestion
	$gparams = json_decode($field->group_params);

	if (isset($field->group_title) && isset($gparams->render_tag))
```

ceford replied:
Fixed - removed trailing white space. I recently installed a new version of Eclipse PDT and did not get the settings right!

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford commented:
Pity this is stuck in limbo. All tests have passed and the adverse comments resulted from testing procedure.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31408">issues.joomla.org/tracker/joomla-cms/31408</a>.</sub>

bembelimen commented:
Nice feature, for sure something for 4.1

ceford commented:
Why put this off for 4.1? It was always intended for 4.0.

infograf768 commented on components/com_fields/layouts/fields/render.php:
```diff
@@ -69,14 +73,68 @@
+		$title_tag[$field->group_title] = $gparams->title_tag;
+		$title_class[$field->group_title] = $gparams->title_class;
+		$output[$field->group_title][] = '<' . $tag  . ' class="field-entry ' . $class . '">' . $content . '</' . $tag  . '>';
+	} else {
```
}
else
{

infograf768 commented on components/com_fields/layouts/fields/render.php:
```diff
@@ -69,14 +73,68 @@
+		$class = empty($title_class[$title]) ? '' : ' class="' . $title_class[$title] . '"';
+		echo '<' . $title_tag[$title] . $class . '>' . $title . '</' . $title_tag[$title] . '>' . "\n";
+	}
+	if (!($list_type == 'p' || $list_type == 'div'))
```
add blank line before if

infograf768 commented on components/com_fields/layouts/fields/render.php:
```diff
@@ -69,14 +73,68 @@
+		$class = empty($list_type_class[$title]) ? '' : ' class="' . $list_type_class[$title]  . '"';
+		echo '<' . $list_type . $class . '>';
+	}
+	if ($list_type == 'dl')
```
add blank line before if

infograf768 commented on components/com_fields/layouts/fields/render.php:
```diff
@@ -69,14 +73,68 @@
+		echo '<dt'. $class . '>' . $title . '</dt>' . "\n";
+	}
+	echo implode("\n", $output[$title]);
+	if (!($list_type == 'p' || $list_type == 'div'))
```
add blank line before if

richard67 commented:
@wilsonge Your opinion on this? 4.0 or 4.1?

ceford commented:
This came up as a gripe in the forum because switching custom fields from dl to ul in J4 ruined perfectly good article appearance. Put it off to 4.1 and expect a lot more gripes!

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

brianteeman commented:
That gripe was already resolved. This adds new functionality

bembelimen commented:
> 
> 
> That gripe was already resolved. This adds new functionality

Exactly. It's a really good enhancement but we should stop adding new stuff into 4.0 to get it released. But with a bit of clean up, we'll have it in 4.1.

ceford committed:
Coming up to date

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

zero-24 committed:
Joomla! 3.10.3 Stable

zero-24 committed:
reset to dev

Commit:
s/overwritting/overwriting (#35859)

richard67 committed:
Fix double translation of groupText for eos quickicon messages (#35844)

alikon committed:
[3][com_contact] - contact webpage text (#35835)

* Update default_address.php

* Update view.html.php

* Update en-GB.com_contact.ini

* Website

zero-24 committed:
[3][com_tags] Created Date cannot change on every save (#35846)

Backports #35724 into 3.10 thanks @alikon
Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

github-actions[bot] committed:
[create-pull-request] automated change (#35955)

Co-authored-by: wilsonge <wilsonge@users.noreply.github.com>

janschoenherr committed:
[3.0] Remove duplicate attributes (#35936)

Fedik committed:
Re-init tinymce when sorting happen, fix #24343 (#34808)

* Re-init tinymce when sorting

* minify

Co-authored-by: Tobias Zulauf <zero-24@users.noreply.github.com>

zero-24 committed:
Make sure only non blocked users can recive actionlogs emails (#35591)

* Make sure only non blocked users can reciver actionlogs emails

* Update administrator/components/com_actionlogs/models/actionlog.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

wilsonge committed:
Add a proper error message if a pack fails to download (#35841)

richard67 committed:
Use html instead of cmd filter for language names in the extensions installer (#35980)

Denitz committed:
fix (#35528)

PhilETaylor commented on components/com_fields/layouts/fields/render.php:
```diff
@@ -69,14 +73,73 @@
+
+	if (isset($field->group_title) && isset($gparams->render_tag))
+	{
+		switch ($gparams->render_tag)
+		{
+			case 'dl':
+			{
+				$tag = 'dd';
+				break;
+			}
+			case 'p':
+			case 'div':
+			{
+				$tag = $gparams->render_tag;
+				break;
+			}
+			default:
+				$tag = 'li';
+		}
```
```suggestion
		switch ($gparams->render_tag)
		{
			case 'dl':
				$tag = 'dd';
				break;
			case 'p':
			case 'div':
				$tag = $gparams->render_tag;
				break;
			default:
				$tag = 'li';
		}
```

PhilETaylor commented on components/com_fields/layouts/fields/render.php:
```diff
@@ -50,6 +50,10 @@
 }
 
 $output = array();
+$list_type = array();
+$show_title = array();
+$title_tag = array();
+$title_class = array();
```
Its 2021.

```suggestion
$output = [];
$list_type = [];
$show_title = [];
$title_tag = [];
$title_class = [];
```

PhilETaylor commented on components/com_fields/layouts/fields/render.php:
```diff
@@ -69,14 +73,73 @@
+		echo '<' . $title_tag[$title] . $class . '>' . $title . '</' . $title_tag[$title] . '>' . "\n";
+	}
+
+	if (!($list_type == 'p' || $list_type == 'div'))
```
```suggestion
	if (!($list_type === 'p' || $list_type === 'div'))
```

PhilETaylor commented on components/com_fields/layouts/fields/render.php:
```diff
@@ -69,14 +73,73 @@
+		echo '<' . $list_type . $class . '>';
+	}
+
+	if ($list_type == 'dl')
```
```suggestion
	if ($list_type === 'dl')
```

PhilETaylor commented on components/com_fields/layouts/fields/render.php:
```diff
@@ -69,14 +73,73 @@
+	}
+	echo implode("\n", $output[$title]);
+
+	if (!($list_type == 'p' || $list_type == 'div'))
```
```suggestion
	if (!($list_type === 'p' || $list_type === 'div'))
```

Commit:
[3] FOF encryption fails on PHP 8 (#35485)

* FOF encryption fails on PHP 8

* refactor so isSupported is only called once in modern times

* Update libraries/fof/encrypt/aes.php

Co-authored-by: Quy <quy@fluxbb.org>

Co-authored-by: Quy <quy@fluxbb.org>
Co-authored-by: Tobias Zulauf <zero-24@users.noreply.github.com>

Krshivam25 commented:
@ceford Can you please commit this changes so we can test this again.

ceford commented:
Any chance someone else can pick this up? Right now I don't know what to do - too long ago for may failing memory.

Commit:
s/Reset/ResetModel (#36125)

Commit:
Namespace class (#36126)

Commit:
fix undefined class in typehint (#36127)

zero-24 committed:
[3.10] Backport php8.1 compatibility fixes (#36083)

* backport php8.1 combat fixes

* missing checks

* Update Table.php

* Update Table.php

* Update Table.php

* Update Table.php

* Update JInstallerAdapterTest.php

* Update JInstallerAdapterTest.php

* Update JInstallerAdapterTest.php

* Update JInstallerAdapterTest.php

* Update Table.php

* Update Table.php

* Update JInstallerAdapterTest.php

* Update JInstallerAdapterTest.php

* Update Table.php

* Update Table.php

* Update JInstallerAdapterTest.php

zero-24 committed:
[3.10] Add a check whether we are running the core backend template and issue an info when that is not the case (#35107)

* add a check whether we are running the core tempaltes where the backend it required and the frontend is recommended

* some more updates

* move to the "recommended box"

* update title

* update two more strings

* Apply suggestions from code review

Co-authored-by: Brian Teeman <brian@teeman.net>

* alpha sort

* add hathor and beez3

* Update administrator/components/com_joomlaupdate/models/default.php

* only do backend template use the proposed message

* remove frontend check and move backend check down the line

* make drone happy

* move it to a top message only

* that should not be here

* move variable

* add link to styles as requested

* force it to show the admin styles

* Update administrator/components/com_joomlaupdate/models/default.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update administrator/components/com_joomlaupdate/models/default.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Apply suggestions from code review

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update administrator/components/com_joomlaupdate/views/default/view.html.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

* Update administrator/components/com_joomlaupdate/models/default.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

* Update administrator/language/en-GB/en-GB.com_joomlaupdate.ini

* i hate you php 5.3...

* remove additional tabs

Co-authored-by: Brian Teeman <brian@teeman.net>
Co-authored-by: Richard Fath <richard67@users.noreply.github.com>
Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

zero-24 committed:
fix since tags

brianteeman committed:
[4.0] template icon (#36136)

Use the correct icon for Templates and Template views instead of the icon for styles

brianteeman committed:
[4.0] spaces to tabs (#36138)

Simple codestyle change from spaces to tabs

brianteeman committed:
Cassopeia rtl update (#36040)

brianteeman committed:
[4.0] remove unused atum rtl css (#36039)

brianteeman committed:
[4.0] swticher css - rtl (#36036)

Updates the css to use logical properties instead of maintaining LTR and RTL. There is no visible difference

brianteeman committed:
[4.0] com_associations rtl (#36035)

Change to use logical properties and remove some duplication

It does exactly the same thing but by using css logical properties we avoid the need to maintain both an LTR and an RTL version

brianteeman committed:
[4.0] header.css unused (#36032)

To the best of my knowledge and testing ability these classes are no longer used and can be safely removed. It looks like they were missed during the numerous refactoring of the atum template.

When testing either confirm they are not used by code review and/or check that the ... menu is still working as desired on smaller screens

brianteeman committed:
[4.0] Atum alert styling (#36031)

This is a replacement to part of the merged PR #30294

It does exactly the same thing but by using css logical properties we avoid the need to maintain both an LTR and an RTL version

There is no visual change.

brianteeman committed:
[4.0] Template tree rtl (#36029)

This is a replacement to the merged PR #6367

It does exactly the same thing but by using css logical properties we avoid the need to maintain both an LTR and an RTL version

There is no visual change.

brianteeman committed:
[4.0] Fancy selectbox fix (#36026)

* [4.0] Fancy selectbox fix

This is a replacement to the merged PR #30739

It does exactly the same thing but by using css logical properties we avoid the need to maintain both an LTR and an RTL version

There is no visual change.

* same for cassiopeia

brianteeman committed:
[4.0] RTL: Correcting sidebar menu (#36025)

This is a replacement to the merged PR #34869

It does exactly the same thing but by using css logical properties we avoid the need to maintain both an LTR and an RTL version[4.0]

There is no visual change. There is no shorthand for border-radius logical properties

brianteeman committed:
[4.0] Atum content quicklinks (#36030)

This is a partial replacement to part of the merged PR #32336

It does exactly the same thing but by using css logical properties we avoid the need to maintain both an LTR and an RTL version

There is no visual change.

janschoenherr committed:
Update FieldsModel.php (#35924)

Fix missing filter in store id

khu5h1 committed:
Solved -- To Top button overlaps information (#35996)

* Solved -- To Top button overlaps information

* Solved -- To Top button overlaps information

chmst committed:
Notification subject: Use Item title, not model name (#35960)

chmst committed:
[4] Avoid the EOL underline of the title link (#35514)

* avoid the EOL underline of the title link

* Solution with htmlhelper

* Update modules/mod_articles_category/tmpl/default_items.php

Co-authored-by: Quy <quy@fluxbb.org>

Co-authored-by: George Wilson <georgejameswilson@googlemail.com>
Co-authored-by: Quy <quy@fluxbb.org>

rjharishabh committed:
[4.0] Versions Toolbar when button disabled (#35761)

PhocaCz committed:
Cassiopeia template - changing Tagline to Tagline - Optional text to show as a subheading parameter - ALTERNATIVE (35720) (#35729)

Denitz committed:
Finder plugins modify component params registry (#35855)

chmst committed:
[4.0] Add information to the error message (#35910)

* Add information to the error message

* remove '.'

* Update administrator/language/en-GB/plg_sampledata_blog.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

Co-authored-by: Brian Teeman <brian@teeman.net>

zero-24 committed:
add clas variable a fix wrong s&r chagnes

wilsonge committed:
Merge branch '3.10-dev' into 4.0-dev

richard67 committed:
Update deleted files from 4.0.4 to 4.0.5 (#36142)

wilsonge committed:
Merge branch '3.10-dev' into 4.0-dev

wilsonge committed:
Tag 4.0.5 RC

wilsonge committed:
Add flag to disable patch packages for now

wilsonge committed:
Reset to dev

wilsonge committed:
Fix codestyle

wilsonge committed:
Improve variable naming

ceford committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

ceford committed:
Trying to regenerate field_group

ceford commented:
I have made a complete pig's ear of the changes requested by Phil. My local repo is in a mess and I don't know how to fix it or revoke my latest commit. Annoyed with myself and ...

ceford committed:
Whiate space in FieldsModel

ceford commented:
I am going to have to redo this from scratch starting from a new repo clone.
